src/matchbox/server/postgresql/utils/insert.py

```diff
@@ -207,67 +207,59 @@ def _cluster_results_to_hierarchical(
-    for parent, child, prob in hierarchy:
-        adj_dict[child].add((parent, prob))
+    hierarchy: list[tuple[bytes, bytes, float]] = []
+    ultimate_parents: dict[bytes, set[bytes]] = defaultdict(set)
 
-    # Process each threshold level, getting clusters at each threshold
+    # Process each threshold level
     for threshold in thresholds:
         threshold_float = float(threshold)
 
-        current_clusters = cluster_df[cluster_df["threshold"] == threshold]
+        # Add new pairwise relationships at this threshold
+        current_probs = prob_df[prob_df["probability"] == threshold_float]
+
+        for _, row in current_probs.iterrows():
+            parent = row["hash"]
+            left_id = row["left_id"]
+            right_id = row["right_id"]
+
+            hierarchy.extend(
+                [
+                    (parent, left_id, threshold_float),
+                    (parent, right_id, threshold_float),
+                ]
+            )
+
+            ultimate_parents[left_id].add(parent)
+            ultimate_parents[right_id].add(parent)
+
+        # Process clusters at this threshold
+        current_clusters = cluster_df[cluster_df["threshold"] == threshold_float]
 
-        # Group by parent to process each component
+        # Group by parent to process components together
         for parent, group in current_clusters.groupby("parent"):
-            members = set(group["child"])
-            if len(members) <= 2:
-                continue
-
-            seen = set(members)
-            current = set(members)
-            ultimate_parents = set()
-
-            # Keep traversing until we've explored all paths
-            while current:
-                next_level = set()
-                # If any current nodes have no parents above threshold,
-                # they are ultimate parents for this threshold
-                for node in current:
-                    parents = {
-                        p for p, prob in adj_dict[node] if prob >= threshold_float
-                    }
-                    next_parents = parents - seen
-                    if not parents:  # No parents = ultimate parent
-                        ultimate_parents.add(node)
-
-                    next_level.update(next_parents)
-                    seen.update(parents)
-
-                current = next_level
-
-            for up in ultimate_parents:
+            children = set(group["child"])
+            if len(children) <= 2:
+                continue  # Skip pairs already handled by pairwise probabilities
+
+            current_ultimate_parents: set[bytes] = set()
+            for child in children:
+                current_ultimate_parents.update(ultimate_parents[child])
+
+            for up in current_ultimate_parents:
                 hierarchy.append((parent, up, threshold_float))
-                adj_dict[up].add((parent, threshold_float))
 
+            for child in children:
+                ultimate_parents[child] = {parent}
```

Comment:
Clever - so unlike in my original suggestion, you're allowing multiple ultimate parents for each node: they are added at the point of adding new 2-node clusters because it's the only way this situation can arise. However, once we merge everything under a single cluster we're again in the world were there must be only one ultimate parent.

I still think there is no reason, [in this example](https://github.com/uktrade/matchbox/pull/20#discussion_r1871161117), to create an edge between `cvtu` and `zv`. But this will have a negligible impact on time now. It has more of an impact on space. I again think there is no good reason to create that edge on the basis that the hierarchical representation needs to be able to reconstruct:
- the probabilities, which it always can because we're storing 2-node clusters at the beginning of the process
- the clusters at any probability threshold, which it can do, because the only requirement is that there exists a path from a point in the hierarchy to all the leaves, which is the case even without the redundancy